### PR TITLE
Evolution bindings

### PR DESCRIPTION
**Pull request by hernot**
Sorry for closing the other pull request and reopening the new one, still learning git and just right now figured how rebase can be used to merge multiple commits, and remove unwanted commits to get things cleaner.

Ok as requested i have made all changes. 

Hope you enjoyed your holidays

Hernot

### CODE REVIEW

**Comment by hernot** [comment]
Hi Philip 
Am Donnerstag, den 21.09.2017, 02:04 -0700 schrieb Philipp Schüle:
> pschuele commented on this pull request.
> 
> hi,
> 
> thanks again for your pull request. i'll put this on our ci server
> for automated testing. if that succeeds, we can finally merge this :)
> btw: did you already sign our CLA? 
> https://wiki.tine20.org/Developers/Getting_Started/Contributors_Guide
> /Contributor_Licence_Agreement
> 


Shure I did. 

Never the less i decided to go via github as my testing setup is not
fully conforming to yours for the following reasons.

1) My system where the server with the test and debugging setup is
remotely located and thus https, ssh access is mandatory

2) The version of the nodejs testing and debugging webserver you use at
your site was not available any more. I had to get a new major
version. 

On GitHub i can push the changes which are related to testing
environment only into a separate subbranch and just pullrequest the
rest.


Xristoph

-- 
Christoph, Elisabeth Hintermüller
Eisenhandstraße 33/23
4020 Linz
Austria
Tel.: +43 650 8827347
mail: christoph@out-world.com
mail: elisabeth@out-world.com
www: http://www.out-world.com

**Comment by pschuele** [comment]
hi,
ah, you already signed, great. I'll check why this isn't reflected in our list.

i have the results of the automated tests:


There was 1 error:

1) Addressbook_Frontend_WebDAV_ContactTest::testCreateContactWithPhoto
Addressbook_Exception_NotFound: Contact has no image.

/usr/local/share/tine20.git/tine20/Addressbook/Controller.php:157
/usr/local/share/tine20.git/tine20/Tinebase/Controller.php:545
/usr/local/share/tine20.git/tine20/Addressbook/Model/Contact.php:549
/usr/local/share/tine20.git/tests/tine20/Addressbook/Frontend/WebDAV/ContactTest.php:107

--


There were 2 failures:

1) Addressbook_Convert_Contact_VCard_MacOSXTest::testConvertToTine20Model
Failed asserting that null matches expected 'Lars'.

/usr/local/share/tine20.git/tests/tine20/Addressbook/Convert/Contact/VCard/MacOSXTest.php:109

2) Addressbook_Convert_Contact_VCard_MacOSXTest::testConvertToVCard
Failed asserting that null matches expected 'Lars'.

/usr/local/share/tine20.git/tests/tine20/Addressbook/Convert/Contact/VCard/MacOSXTest.php:109
/usr/local/share/tine20.git/tests/tine20/Addressbook/Convert/Contact/VCard/MacOSXTest.php:116

FAILURES!
Tests: 2253, Assertions: 13002, Failures: 2, Errors: 1, Incomplete: 16, Skipped: 105.

---
would you like to check those tests on your system? we could also try to fix this but this may take a while...

cu
Philipp

**Comment by hernot** [comment]
As i already stated in my private email i'm quite busy to get my researchgroup going, so i would appriciate if you could have a look into. As i think it will take me longer than you to find a large enough spare time slot to fix.

**Comment by pschuele** [comment]
should be rebased on the current master. maybe the failing tests could also be added to https://github.com/tine20/Tine-2.0-Open-Source-Groupware-and-CRM/blob/master/tests/tine20/TravisTests.php

**Comment by hernot** [comment]
Hi Any
Do you still need anything from my side or may i consider this fully handled by you?

**Comment by pschuele** [comment]
hi Hernot,
it would be great, if you could do 2 things:
* sign the cla (via the assistant)
* rebase the change on the current master

... i will add the tests so that travis shows the failing ones.

**Comment by msphn** [comment]
vcard changes not tested by travis, atm only json tests there

**Comment by pschuele** [comment]
> vcard changes not tested by travis, atm only json tests there

those tests could be added to the travis test suite in the PR.

**Comment by hernot** [comment]
Hi Philipp

Inline Response (answers below questions asked)
Message sent on Freitag, 18.05.2018, 01:48 -0700 by Philipp Schüle:
> hi Hernot,
> it would be great, if you could do 2 things:
> * sign the cla (via the assistant)
> * rebase the change on the current master
> 


Just to be on the same page: By rebase on the current master you mean.

Pull the latest version of the master branch to my github and local
repos. And than merge the new master onto the feature branch, test if
anything is broken and push back to my githup repo for updating the
pullrequest?


Best Xristoph

**Comment by hernot** [comment]
Hi Philipp

Inline Response (answers below questions asked)
Message sent on Freitag, 18.05.2018, 01:48 -0700 by Philipp Schüle:
> hi Hernot,
> it would be great, if you could do 2 things:
> * sign the cla (via the assistant)
> * rebase the change on the current master
> 



When will be the deadline for the release and till when this should be
done, to be still considered? Just for planning my time properly.

Best 
Xristoph

-- 
Christoph, Julia, R. Hintermüller
Eisenhandstraße 33/23
4020 Linz
Austria
Tel.: +43 650 8827347
mail: christoph@out-world.com
mail: julia@out-world.com
www: http://www.out-world.com

**Comment by pschuele** [comment]
hi Xristoph,

> Just to be on the same page: By rebase on the current master you mean.
Pull the latest version of the master branch to my github and local
repos. And than merge the new master onto the feature branch, test if
anything is broken and push back to my githup repo for updating the
pullrequest?

here's how to rebase your pull request: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request (note: you don't merge the master into your feature branch but rebase the feature branch on the current master). the PR needs to be updated with a force push. you also should squash the commits - a single commit is preferred (we'll add that information to the contributing guidelines soon).

> When will be the deadline for the release and till when this should be
done, to be still considered?

the next release is scheduled for the 21st of June (see https://github.com/tine20/Tine-2.0-Open-Source-Groupware-and-CRM/milestone/2).

cheers!
Philipp

**Comment by pschuele** [comment]
added a new PR with the activated Addressbook tests that failed with this PR: #79 
maybe you could add this commit to your PR as well to get those tests running: https://github.com/tine20/Tine-2.0-Open-Source-Groupware-and-CRM/pull/79/commits/9a91599e90784254e6d64be1dacdd7271140adef

**Comment by pschuele** [comment]
you see the failing tests here (and how to run the tests): https://travis-ci.org/tine20/Tine-2.0-Open-Source-Groupware-and-CRM/builds/386535321

**Comment by hernot** [comment]
OK Thanks will take care. 

LG
Xristoph
Inline Response (answers below questions asked)
Message sent on Freitag, 01.06.2018, 00:36 -0700 by Philipp Schüle:
> added a new PR with the activated Addressbook tests that failed with
> this PR: #79 
> maybe you could add this commit to your PR as well to get those tests
> running: https://github.com/tine20/Tine-2.0-Open-Source-Groupware-and
> -CRM/pull/79/commits/9a91599e90784254e6d64be1dacdd7271140adef
>

File: tine20/Addressbook/Convert/Contact/VCard/Abstract.php [hunk 1]
```diff
@@ -176,10 +176,73 @@ public function toTine20Model($blob, Tinebase_Record_Abstract $_record = null, $
...
+                         && ( $property['ENCODING'] != "B" ) ) {
+                        break;
+                    }
+                    switch ( $property['TYPE'] ) {
+                        case 'JPG' : {}
+                        case 'jpg' : {}
+                        case 'Jpg' : {}
+                        case 'Jpeg' : {}
+                        case 'jpeg' : {}
+			case 'PNG' : {}
```
Comment (msphn) [inline tine20/Addressbook/Convert/Contact/VCard/Abstract.php]: indentation

File: tine20/Addressbook/Convert/Contact/VCard/Abstract.php [hunk 2]
```diff
@@ -176,10 +176,73 @@ public function toTine20Model($blob, Tinebase_Record_Abstract $_record = null, $
...
+                            if (Tinebase_Core::isLogLevel(Zend_Log::DEBUG)) 
+                                Tinebase_Core::getLogger()->debug(__METHOD__ . '::' . __LINE__ . ' Photo: passing on jpeg image as is (' . strlen($property->getValue()) .')' );
+                            $jpegphoto = $property->getValue();
+                            break;
+                        }
+                        default : {
+                            if (Tinebase_Core::isLogLevel(Zend_Log::DEBUG)) 
+                                Tinebase_Core::getLogger()->debug(__METHOD__ . '::' . __LINE__ . ' Photo: recoding to jpeg (' . strlen($property->getValue()) . ')' );
+/*
+			    $binimage = imagecreatefromstring($property->getValue());
```
Comment (msphn) [inline tine20/Addressbook/Convert/Contact/VCard/Abstract.php]: if you comment it, remove it

File: tine20/Addressbook/Convert/Contact/VCard/Abstract.php [hunk 3]
```diff
@@ -176,10 +176,73 @@ public function toTine20Model($blob, Tinebase_Record_Abstract $_record = null, $
...
+                                    Tinebase_Core::getLogger()->debug(__METHOD__ . '::' . __LINE__ . ' Photo: not recoded ' );
+                            	$jpegphoto = $property->getValue();
+                            	break;
+                            }
+                            $recode = imagecreatefromstring($property->getValue());
+                            if ( $recode === false ) {
+                                break;
+                            }                            
+                            ob_start();
+                            if ( imagejpeg($recode,null,90) === true ) {
```
Comment (msphn) [inline tine20/Addressbook/Convert/Contact/VCard/Abstract.php]: PSR-1, PSR-2 please!

File: tine20/Addressbook/Convert/Contact/VCard/Abstract.php [hunk 4]
```diff
@@ -208,8 +271,9 @@ public function toTine20Model($blob, Tinebase_Record_Abstract $_record = null, $
                     break;
                     
                 default:
-                    if (Tinebase_Core::isLogLevel(Zend_Log::DEBUG)) 
-                        Tinebase_Core::getLogger()->debug(__METHOD__ . '::' . __LINE__ . ' cardData ' . $property->name);
+                    $this->_toTine20ModelParseOther($data,$property);
+                    /*if (Tinebase_Core::isLogLevel(Zend_Log::DEBUG))
```
Comment (msphn) [inline tine20/Addressbook/Convert/Contact/VCard/Abstract.php]: remove

File: tine20/Addressbook/Convert/Contact/VCard/Abstract.php [hunk 5]
```diff
@@ -448,18 +529,23 @@ protected function _fromTine20ModelAddPhoto(Addressbook_Model_Contact $record, \
      * @param  Tinebase_Record_Abstract  $record
      * @return \Sabre\VObject\Component
      */
-    protected function _fromTine20ModelRequiredFields(Tinebase_Record_Abstract $record)
+    protected function _fromTine20ModelRequiredFields(Tinebase_Record_Abstract $record,$fn = null,$org = null)
```
Comment (msphn) [inline tine20/Addressbook/Convert/Contact/VCard/Abstract.php]: code styling here again

File: tine20/Addressbook/Convert/Contact/VCard/Abstract.php [hunk 6]
```diff
@@ -448,18 +529,23 @@ protected function _fromTine20ModelAddPhoto(Addressbook_Model_Contact $record, \
...
     {
         /** @var Addressbook_Model_Contact $record */
         $version = Tinebase_Application::getInstance()->getApplicationByName('Addressbook')->version;
-        
+       	if ( !isset($fn) || is_null($fn) ) {
+            $fn = $record->n_fileas;
+        }
+       	if ( !isset($org) || is_null($org) ) {
```
Comment (msphn) [inline tine20/Addressbook/Convert/Contact/VCard/Abstract.php]: don't use is_null, use null === $org and $fn!

File: tine20/Addressbook/Convert/Contact/VCard/Evolution.php [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        'assistent'             => null,
+        'bday'                  => null,
+        'calendar_uri'          => null,
+        'email'                 => null,
+        'email_home'            => null,
+        'jpegphoto'             => null,
+        'freebusy_uri'          => null,
+        'note'                  => null,
+        'role'                  => null,
+        #'salutation'            => null,
```
Comment (msphn) [inline tine20/Addressbook/Convert/Contact/VCard/Evolution.php]: don't leave commented code please

File: tine20/Addressbook/Convert/Contact/VCard/Abstract.php (hunk 3 above)
Comment (hernot) [reply tine20/Addressbook/Convert/Contact/VCard/Abstract.php]: Just a quick question, psr-1 psr-2 what do you mean by that or are they implicitly applied if i rebase my branche to latest head or will that show by the Tarvis CI anyway?

**Comment by hernot** [comment]
Ok the two failing tests are about image beeing not a valid image which could be recoded if necessary to a image format and/or size supported by tine20.

Currently no photo is added at all if image blob is not a valid image format or can not be recoded if necessary. What would be the default policy to apply in these cases?

**Comment by hernot** [comment]
Quick question and do not know if if find still enough time till thursday 21st but by looking at the/usr/share/tine20debug/tests/tine20/Addressbook/Convert/Contact/VCard/EMClientTest.php File if should i add a tests for the evoulution bindings too ? or should i postpone that to the next release cycle to be on the safe side?

**Comment by msphn** [comment]
You are very welcome to add a test! I'll wait for it and review as soon as you ping me and your tests are green!

**Comment by pschuele** [comment]
> And about my other questions, related to expected default behavior/policy when photo is missing, or encoded in a format which is not supported by tine20 and can not with standard php packages methods be converted into appropriate format and size or is just a simple text as in your MacOSx test. Is that a case which should be valid and what for reason?

I think there is a problem with the image (PHOTO) handling in your change. the abstract converter should still be able to handle VCARD PHOTOS like the one in https://github.com/tine20/tine20/blob/master/tests/tine20/Addressbook/Import/files/jan.vcf (which does not have a TYPE attribute). maybe you should move the new PHOTO handling to your custom Evolution converter?

**Comment by msphn** [comment]
```1) Addressbook_Frontend_WebDAV_ContactTest::testCreateContactWithPhoto
Addressbook_Exception_NotFound: Contact has no image.
/home/travis/build/tine20/tine20/tine20/Addressbook/Controller.php:157
/home/travis/build/tine20/tine20/tine20/Tinebase/Controller.php:577
/home/travis/build/tine20/tine20/tine20/Addressbook/Model/Contact.php:862
/home/travis/build/tine20/tine20/tests/tine20/Addressbook/Frontend/WebDAV/ContactTest.php:108
--
There were 2 failures:
1) Addressbook_Convert_Contact_VCard_MacOSXTest::testConvertToTine20Model
Failed asserting that null matches expected 'Lars'.
/home/travis/build/tine20/tine20/tests/tine20/Addressbook/Convert/Contact/VCard/MacOSXTest.php:105
2) Addressbook_Convert_Contact_VCard_MacOSXTest::testConvertToVCard
Failed asserting that null matches expected 'Lars'.
/home/travis/build/tine20/tine20/tests/tine20/Addressbook/Convert/Contact/VCard/MacOSXTest.php:105
/home/travis/build/tine20/tine20/tests/tine20/Addressbook/Convert/Contact/VCard/MacOSXTest.php:112

```

**Comment by hernot** [comment]
Ok will squash my commits so far to have a clean commit history. When lastest will you close this pullrequest for merging into thursday release or should i create a new pull request after 21st of june for contributing the evolution unit tests.

**Comment by pschuele** [comment]
great - the tests are running :+1: 

> Asking cause i will not start today to implement the tests for evolution converter and not sure whether i manage till thursday. Would it be a better solution when i create a new pull request after 21st of june for contributing the evolution unit tests.

no problem, just open a new PR with the test(s).

@msphn please do a final review :)

**Comment by pschuele** [comment]
thx @hernot for this pr :smiley:

**Pull request merged**